lib/finch/http1/pool.ex

```diff
@@ -38,6 +38,158 @@ defmodule Finch.HTTP1.Pool do
     )
   end
 
+  @impl Finch.Pool
+  def stream(pool, req, name, opts) do
+    pool_timeout = Keyword.get(opts, :pool_timeout, 5_000)
+    receive_timeout = Keyword.get(opts, :receive_timeout, 15_000)
+    request_timeout = Keyword.get(opts, :request_timeout, 30_000)
+    fail_safe_timeout = Keyword.get(opts, :fail_safe_timeout, 15 * 60_000)
+    stop_notify = Keyword.get(opts, :stop_notify, nil)
+
+    metadata = %{request: req, pool: pool, name: name}
+
+    start_time = Telemetry.start(:queue, metadata)
+    owner = self()
+    ref = make_ref()
+
+    holder =
+      spawn_link(fn ->
+        try do
+          NimblePool.checkout!(
+            pool,
+            :checkout,
+            fn from, {state, conn, idle_time} ->
+              if fail_safe_timeout != :infinity do
+                Process.send_after(self(), :fail_safe_timeout, fail_safe_timeout)
+              end
+
+              Telemetry.stop(:queue, start_time, metadata, %{idle_time: idle_time})
+
+              return =
+                case Conn.connect(conn, name) do
+                  {:ok, conn} ->
+                    send(owner, {ref, :ok, {conn, idle_time}})
+
+                    receive do
+                      {^ref, :stop, conn} ->
+                        with :closed <- transfer_if_open(conn, state, from) do
+                          {:ok, :closed}
+                        end
+
+                      :fail_safe_timeout ->
+                        Conn.close(conn)
+                        {:fail_safe_timeout, :closed}
+                    end
+
+                  {:error, conn, error} ->
+                    {{:error, error}, transfer_if_open(conn, state, from)}
+                end
+
+              with {to, message} <- stop_notify do
+                send(to, message)
+              end
+
+              return
+            end,
+            pool_timeout
+          )
+        else
+          :fail_safe_timeout ->
+            raise "Fail safe timeout was hit"
+
+          other ->
+            other
+        catch
+          :exit, data ->
+            Telemetry.exception(:queue, start_time, :exit, data, __STACKTRACE__, metadata)
+            send(owner, {ref, :exit, {data, __STACKTRACE__}})
+        end
+      end)
+
+    receive do
+      {^ref, :ok, conn_idle_time} ->
+        {conn, idle_time} = conn_idle_time
+
+        stream =
+          fn
+            {:cont, acc}, function ->
+              case Process.alive?(holder) do
+                true ->
+                  Process.link(holder)
+
+                false ->
+                  raise "Process holding the connection in pool died" <>
+                          " before the stream enumeration started." <>
+                          " Most likely fail_safe_timeout occured"
+              end
+
+              try do
```

Comment:
The current code is not dealing with stream suspensions. Try implementing `Stream.zip(stream, stream)`. It needs to spawn two separate connections (instead of using the same one) and be able to yield out of this loop and come back in.

Reply:
> The current code is not dealing with stream suspensions

Yeah, I don't understand what suspension means, that's why I decided to use `Stream.resourse` in the first place. I've read `Enum` and `Enumerable` documentation and haven't found any explanation of what halt and suspend mean and how to treat them

Reply:
Oh, I've found it, it's in type docs. I'll try to use it, thanks